bmc/boot_device.go

```diff
@@ -78,3 +95,67 @@ func SetBootDeviceFromInterfaces(ctx context.Context, timeout time.Duration, boo
 	}
 	return setBootDevice(ctx, timeout, bootDevice, setPersistent, efiBoot, bdSetters)
 }
+
+// getBootDeviceOverride gets the boot device override settings for the given provider,
+// and updates the given metadata with provider attempts and errors.
+func getBootDeviceOverride(
+	ctx context.Context,
+	timeout time.Duration,
+	provider *bootOverrideProvider,
+	metadata *Metadata,
+) (override *BootDeviceOverride, err error) {
+	select {
+	case <-ctx.Done():
+		err = multierror.Append(err, ctx.Err())
+		return nil, err
+	default:
+		metadata.ProvidersAttempted = append(metadata.ProvidersAttempted, provider.name)
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+
+		override, err = provider.bootOverrider.BootDeviceOverrideGet(ctx)
+		if err != nil {
+			metadata.FailedProviderDetail[provider.name] = err.Error()
+			return nil, nil
+		}
+
+		metadata.SuccessfulProvider = provider.name
+		return override, nil
+	}
+}
+
+// GetBootDeviceOverrideFromInterface will get boot device override settings from the first successful
+// call to a BootDeviceOverrideGetter in the array of providers.
+func GetBootDeviceOverrideFromInterface(
+	ctx context.Context,
+	timeout time.Duration,
+	providers []interface{},
+) (*BootDeviceOverride, Metadata, error) {
+	var err error
+	metadata := Metadata{
+		FailedProviderDetail: make(map[string]string),
+	}
+
+	for _, elem := range providers {
+		provider := &bootOverrideProvider{name: getProviderName(elem)}
+		switch p := elem.(type) {
+		case BootDeviceOverrideGetter:
+			provider.bootOverrider = p
+			override, getErr := getBootDeviceOverride(ctx, timeout, provider, &metadata)
```

Comment:
For consistency in these interface identifier methods, I suggest collecting all the implementations of `BootDeviceOverrideGetters` in a slice and returning `getBootDeviceOverride` at the end, similar to https://github.com/bmc-toolbox/bmclib/blob/b0ef181d6ed441ea5bd9e7161e92712633a28811/bmc/boot_device.go#L42 
Unless you have some reason to not follow that pattern here

Reply:
I did notice that implementation. I decided to implement it differently primarily because `setBootDevice` uses a `defer` inside a `for` loop, which can cause multiple deferred executions to build up until the function returns. I know we don't really have that many things to loop over, but thought I would at least try to do the right thing and move the defer.

Also, I thought it would be more simplistic to loop over the slice of providers once instead of twice (once to filter, then again to do the calls).

Do you still recommend I filter and then return `getBootDeviceOverride` for consistency?

Reply:
Ah, now that I see the reasoning for the difference, it makes sense to me. As long as the behaviour is consistent and new providers can be added without any changes here - this is fine. Thanks!
At some point we need to figure if these boilerplate like methods can be generated along with tests.